src/main/java/net/oauth/jsontoken/AsyncJsonTokenParser.java

```diff
@@ -0,0 +1,148 @@
+   * when getting the result of the future, an {@link ExecutionException} may be thrown
+   * in which {@link ExecutionException#getCause()} follows the same possible exceptions
+   * as thrown by {@link JsonTokenParser#verify(JsonToken)}.
+   *
+   * @param jsonToken
+   * @return a {@link ListenableFuture} that will return a Void
+   */
+  public ListenableFuture<Void> verify(JsonToken jsonToken) {
+    ListenableFuture<List<Verifier>> futureVerifiers = provideVerifiers(jsonToken);
+    // Use AsyncFunction instead of Function to allow for checked exceptions to propagate forward
+    AsyncFunction<List<Verifier>, Void> verifyFunction =
+        verifiers -> Futures.immediateFuture(verifyAndReturnVoid(jsonToken, verifiers));
+
+    return Futures.transformAsync(futureVerifiers, verifyFunction, executor);
+  }
+
+  /**
+   * Parses, and verifies, a JSON Token.
+   * This method is not expected to throw exceptions when returning a future. However,
+   * when getting the result of the future, an {@link ExecutionException} may be thrown
+   * in which {@link ExecutionException#getCause()} follows the same possible exceptions
+   * thrown by {@link JsonTokenParser#verifyAndDeserialize(String)}.
+   *
+   * @param tokenString the serialized token that is to parsed and verified.
+   * @return a {@link ListenableFuture} that will return the deserialized {@link JsonObject},
+   * suitable for passing to the constructor of {@link JsonToken}
+   * or equivalent constructor of {@link JsonToken} subclasses.
+   */
+  public ListenableFuture<JsonToken> verifyAndDeserialize(String tokenString) {
+    ListenableFuture<String> futureTokenString = Futures.immediateFuture(tokenString);
+    AsyncFunction<String, JsonToken> verifyFunction =
+        tokenStringInput -> {
+          JsonToken jsonToken = deserialize(tokenStringInput);
+          Function<Void, JsonToken> returnFunction = unused -> jsonToken;
+          return Futures.transform(verify(jsonToken), returnFunction, executor);
+        };
+
+    return Futures.transformAsync(futureTokenString, verifyFunction, executor);
+  }
+
+  /**
+   * Use {@link AsyncVerifierProviders} to get future that will return a list of verifiers
+   * for this token
+   *
+   * @param jsonToken
+   * @return a {@link ListenableFuture} that will return a list of verifiers
+   */
+  private ListenableFuture<List<Verifier>> provideVerifiers(JsonToken jsonToken) {
+    ListenableFuture<JsonToken> futureJsonToken = Futures.immediateFuture(jsonToken);
+    AsyncFunction<JsonToken, List<Verifier>> findVerifiersFunction =
+        token -> asyncVerifierProviders
+            .getVerifierProvider(token.getSignatureAlgorithm())
+            .findVerifier(token.getIssuer(), token.getKeyId());
+
+    Function<List<Verifier>, List<Verifier>> checkNullFunction =
+        verifiers -> {
+          if (verifiers == null) {
+            throw new IllegalStateException("No valid verifier for issuer: " + jsonToken.getIssuer());
```

Comment:
Where did we land on choice of exception for new code? I don't remember the conclusion of the other thread.

Reply:
We never came up with a specific solution for handling exceptions, as it seems we can't make use of (compiler enforced) checked exceptions with futures. For now, we decided to just create the async interface with the identical exceptions as the sync underneath and add alternative exception handling later when everything is working properly.

Reply:
If we wanted to have better exception handling in this PR, how does this solution sound: 

We have custom exceptions specifically for the async interface (doesn't extend IllegalStateException or any of the other inappropriately thrown exceptions like my previous idea). We can embed extra information in the synchronous exceptions by setting throwable `cause` to one of these custom exceptions. Whenever we call the synchronous methods, we can catch the exceptions and throw the underlying cause in the asynchronous version.

Reply:
That approach sgtm since I don't think any code using the old interface would be broken by adding a cause. Let's do it in a separate PR though since we're already doing a lot in this one.

Reply:
A possible issue with my approach is, even though the library doesn't explicitly use cause, there are some exceptions thrown that do (eg. JsonParseException) or implementations that **might** use it (eg. the Checkers interface and the thrown signature exception). In some instances, we may have to override the existing cause with this method. Would this be a big issue?